utils/crates-io/src/simulator.rs

```diff
@@ -0,0 +1,105 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Packages publishing simulator
+
+use crate::{Workspace, CARGO_REGISTRY_NAME};
+use anyhow::Result;
+use std::{
+    fs,
+    net::SocketAddr,
+    ops::Deref,
+    path::{Path, PathBuf},
+};
+use tempfile::TempDir;
+use tokio::task::{self, JoinHandle};
+use toml_edit::DocumentMut;
+
+enum RegistryPath {
+    Dir(PathBuf),
+    TempDir(TempDir),
+}
+
+impl Deref for RegistryPath {
+    type Target = Path;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            RegistryPath::Dir(path) => path,
+            RegistryPath::TempDir(temp_dir) => temp_dir.path(),
+        }
+    }
+}
+
+/// crates-io packages publishing simulator.
+#[allow(dead_code)]
+pub struct Simulator {
+    path: RegistryPath,
+    addr: SocketAddr,
+    handle: JoinHandle<()>,
+    config_path: PathBuf,
+    original_config: DocumentMut,
+    mutable_config: DocumentMut,
+}
+
+impl Simulator {
+    /// Create a new simulator.
+    pub fn new(registry_path: Option<PathBuf>) -> Result<Self> {
+        let path = match registry_path {
+            Some(path) => RegistryPath::Dir(path),
+            None => RegistryPath::TempDir(TempDir::new()?),
+        };
+        let (future, addr) = cargo_http_registry::serve(&path, "127.0.0.1:35503".parse()?)?;
```

Comment:
pick an available port instead of this fixed one?

Reply:
I wanted to pick a random port, but it seems like that starts creating a lot of folders in `~/.cargo/registry/index` that start with `127.0.0.1-`